### PR TITLE
feat(ci): stop the next leaked temp file, and price the ones already there

### PR DESCRIPTION
**Pull request by WingedGuardian**
## The class

An atomic write is *write to a temp, then rename it into place*. When the write or
the rename raises and nothing unlinks the temp, the temp survives.

Not a hypothesis — `~/.genesis/tmp9cis_fly.tmp`, 0 bytes, on disk since
**2026-05-24** with `mkstemp`'s default naming. One file proves both halves: the
leak happens, and nothing sweeps that directory. `disk_hygiene.sh` roots every
`find` at a named *sub*directory; `tmp_watchgod.sh` covers `~/.genesis/cc-tmp` and
`/tmp`. Neither covers the `~/.genesis` root.

## Guard first, fixes after

MEASURED against the MERGE of this branch into main, 2026-09-09: **60 atomic-write
sites across 52 files, 30 dirty.** Fixing 30 instances of a recurring pattern
leaves nothing to stop instance 31 — so the mechanism ships first, and the 30 are
grandfathered in a debt ledger that can only shrink.

That denominator has moved four times, and each move was a detector blind spot
rather than a code change — which is the argument for the guard, not against it.
It opened at 55/48; anchoring on `mkstemp` missed hand-rolled temp names; an
end-anchored suffix test made an f-string temp invisible; `#1609` landed on main
and deleted a baselined site while this PR was open; and an adversarial audit
found publish-by-hardlink (`os.link`) and module-constant temp markers invisible
too. A count is only ever a census of the shapes the detector can see, so it is
the claim most likely to be wrong — re-derive it rather than quoting this line.
Note the counts come from the MERGE, not the branch: they differ, and CI reads
the merge. The
guard **fails** on a row that no longer matches, so a landed fix cannot leave a
stale entry behind. Polarity is allowlist: neither clean nor baselined ⇒ fail.

## The first ledger was 49 rows at 67% precision, and 16 were booby-trapped

This is the part worth reading. Anchoring on the verb alone was wrong —
`rename`/`replace` also covers move-aside, claim-by-rename, rotate and quarantine,
where the first operand is **durable**.

The guard *prints a remediation*, so a ledger row is a work item someone will act
on. "Unlink the temp", applied to those rows, would have deleted:

| site | what the first operand actually is |
|---|---|
| `guardian/cred_integrity.py:324` | a **live credential** being preserved as `.corrupt-<stamp>` — on a function this same guard already classified `CLEANS_UP` |
| `dashboard/routes/files.py:300` | a **user's file** in the dashboard file manager |
| `observability/span_ingest.py:129` | pending telemetry, on its explicit restore-on-DB-failure path |
| `guardian/alert/queue.py:197` | a corrupt entry quarantined precisely to **preserve the evidence** |

A false row in a debt ledger is worse than noise. Fixed at the mechanism — an
**origin test** for the operand (the temp must be born in the same function), not
a suppression list. That *tightens* the allowlist rather than loosening it.

## Recall was measured in the same breath

A precision fix measured alone is half a measurement. The first tightening **lost**
`autonomy/cli_policy.py`, a known real leak, whose temp is created by
`with NamedTemporaryFile(...) as tmp` and renamed via `tmp_path = Path(tmp.name)` —
two hops, matching no temp-maker on its own. `_born_in` is therefore transitive.

Result: all 9 known leaks caught, all 16 destructive false positives gone. Sites
78 → 55, ledger 49 → 31.

## Also fixed

- **`shutil.move` was invisible** — 5 live sites, one of them three lines above a
  baselined leak in the same function.
- **Baseline keys used a bare function name**, so a clean `A._write` could absorb a
  genuinely new leaking `C._write`. Now `Class.method`, with a collision assertion.
- **A scan that saw zero files exited 0** — `Path.rglob` on a missing directory
  yields nothing rather than raising, so a mis-rooted run was a green check for
  examining nothing. Now a self-maintaining floor.
- **`dataclasses.replace` attribute form** — the one this repo actually uses,
  4 sites — sailed past an exclusion written only for the bare form.
- **`atomic_write_text`'s cleanup branch had no test at all**, and every future fix
  in this class depends on it. Now covered against a real `IsADirectoryError`
  rather than a monkeypatch, plus a test that a failed write leaves the original
  intact.

## Verification

- **The guard bites**: a synthetic leaking module → exit 1 naming file, line and
  function; removed → exit 0.
- **Controls flip both ways**: 3 known-clean read `CLEANS_UP`, 9 known-leaking read
  dirty, 9 durable-operand and dataclasses shapes read as nothing.
- **Mutation-verified eight ways**, each restored with byte-identity confirmed by
  hash.
- 35 targeted tests · `ruff` clean.

**That sweep found two of my own tests vacuous, and they were fixed rather than
explained away.** The key-collision test took its temp as a *parameter*, so
born-here excluded both methods and it asserted `len(set()) == len([])`. The
empty-scan test returned 1 through the *stale* branch and stayed green with the
floor deleted. `main()`'s four outcomes are now driven with controlled inputs and
asserted on each branch's own message.

One more worth recording: the qualified-name fix **silently failed to apply** the
first time — an unasserted string replace whose anchor no longer matched, so the
function was defined and never called. The guard's *own* stale-row check caught
it, reporting 5 NEW and 5 STALE rows differing only by qualification. Every edit
after that asserts its anchor matched exactly once.

## Review

Internal adversarial pass — no cross-model round consumed. Two BLOCKERs, three
SHOULD-FIX, five NOTEs. One reviewer claim was checked and **rejected**: it
reported "0 bare `dataclasses.replace` uses", but 7 files import that form, so the
existing Attribute test was doing real work and was merely incomplete. Acting on
that reasoning would have opened the form it does catch.

E2E: none — a CI guard with no runtime surface. It was executed against the live
tree in both directions before commit (figures above): clean at 55/31/0-new, and
exit 1 naming the site when a synthetic leak is introduced.

## Summary by CodeRabbit

- **Reliability**
  - Improved handling of failed file writes so temporary files are cleaned up and existing files remain unchanged.

- **Quality Improvements**
  - Added automated validation to detect temporary-file leaks during atomic writes.
  - Added baseline tracking and continuous integration checks for write safety.

- **Tests**
  - Expanded coverage for cleanup behavior, detection accuracy, edge cases, scanner failures, and baseline consistency.

### CODE REVIEW

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
**Automated first-pass review** — structural + PII/secrets/wording, both domains run (this PR touches code, not just docs).

### Structural review

No ERROR-level pattern violations found:
- No `nohup`, no direct `task_states` inserts, no `/tmp/` usage (all temp/config paths are repo-relative or generic `~/.genesis`), no `git add -A`/force-push/`reset --hard`, no worktree-path `pip install -e`, no bare `except:` (all handlers are typed — `except (OSError, SyntaxError, ValueError)`, `except (OSError, ValueError, KeyError)`, `except BaseException:` in the atomic-write cleanup path, etc.), no mutable default arguments, no `os.killpg`, no raw SQL against `genesis.db`.

Import hygiene: clean in all three touched `.py` files — stdlib → third-party → local grouping is correct (e.g. `tests/test_util/test_atomic.py` adds `import pytest` before the local `genesis.util.atomic` import), no star or circular imports.

Coverage: the new `scripts/check_atomic_writes.py` has a matching `tests/test_scripts/test_check_atomic_writes.py` with real breadth (shape controls in both directions, process-level checks, and mutation-guarded `main()` branch tests) — exceeds the bar rather than just meeting it.

- **NOTE**: `scan()` (`scripts/check_atomic_writes.py`) and `main()` (same file) are public functions with no docstrings — behavior is well covered by the module-level docstring and inline comments, so this is cosmetic only.

Commit/PR prefix: `feat(ci):` present, conforms to convention. No `IntervalTrigger`/`$CLAUDE_PROJECT_DIR` misuse (not applicable — no scheduling or hook code touched).

### PII/Secrets/Wording

PII/Secrets: clean (no hard blocks, no soft warnings) — no API keys/tokens/passwords, no private IPv4/IPv6 ranges, no SSH keys, no DB connection strings, no personal emails, no usernamed absolute paths, no location-revealing timezones, no cloud account IDs, no hardware identifiers, no private hostnames anywhere in the added lines.

Wording: PR body discusses the guard's own false-positive analysis (e.g. "a live credential being preserved", `guardian/cred_integrity.py`, `dashboard/routes/files.py`) — this is describing the *repo's own source paths* the static analyzer classified, not an actual secrets/PII exposure or private infrastructure (no IPs, hostnames, or real credentials named). Reads as ordinary engineering narrative, not an incident disclosure.

No blocking findings.

VERDICT: PASS




---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by WingedGuardian** [comment]
All three fixed in `44e13e8b3`. The Major one was verified before acting, and the reproduction is worth showing:

```
os.unlink(tmp_backup)  in the handler, temp = tmp   ->  CLEANS_UP
```

on a genuinely leaking site. A false CLEAN is strictly worse than a false flag here — the leak is invisible *and* excluded from the debt ledger, so nothing ever revisits it. An earlier adversarial pass raised this as a NOTE and I accepted it as documented; it showed the mechanism was accidental, you showed it changes the verdict. Accepting it as documented was the error.

Cleanup is now credited by identity, with path-neutral wrappers stripped so `Path(tmp).unlink()` still counts.

**Recall checked in the same breath, because identity alone was also wrong.** It turned the false CLEAN into a false FLAG at `autonomy/executor/engine.py:1824`, where the cleanup *rebuilds* the path (`Path(plan_path).expanduser().with_suffix(".tmp").unlink()`) rather than reusing `tmp`. That site is genuinely clean, so baselining it would have written a false debt row — the exact defect this PR's earlier round removed. Fixed with bounded single-assignment alias resolution, applied recursively; root-only substitution left `path.with_suffix('.tmp')` against the fully-spelled form and they never met.

Net effect on the ledger: 55 sites, 24 clean, 31 dirty, 0 new — identical to before. The verdicts were right; the mechanism producing them wasn't.

Also fixed the `parts` annotation, and set `persist-credentials: false` on **this PR's job only** — the other 14 are out of scope and some may legitimately need the credential.

Both directions pinned, including `.tmpl` contains `.tmp` but is a template (this repo writes `.service.template`, so matching anywhere would have flagged template renders as leaked temps). Mutation-verified three ways; the suffix one survived the first sweep because the original test's RHS had no string literal, so a test that exercises it was added.

**Comment by WingedGuardian** [comment]
@codex review

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -448,6 +448,22 @@ jobs:
...
+  atomic-write-check:
+    # An atomic write that orphans its temp on the exception path leaks a file
+    # into a directory nothing sweeps. MEASURED 2026-09-07: 55 sites, 31 dirty --
+    # and a 0-byte temp had sat in ~/.genesis for 3.5 months, a root no reaper
+    # covers. ALLOWLIST polarity against a shrinking debt ledger, so instance 32
+    # cannot arrive silently while the existing 31 are fixed.
+    # See scripts/check_atomic_writes.py.
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v7
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply .github/workflows/ci.yml:472]: Fixed in `44e13e8b3`.

The substring→identity finding was the important one and I reproduced it before acting: `os.unlink(tmp_backup)` credited cleanup for temp `tmp`, so a genuinely leaking site read `CLEANS_UP`. A false clean is worse than a false flag here — the leak is invisible *and* excluded from the debt ledger, so nothing revisits it. An earlier pass raised this as a NOTE and I accepted it as documented; that was the error.

Recall was checked in the same breath, because identity alone was also wrong: it turned the false clean into a false flag at `engine.py:1824`, where the cleanup rebuilds the path rather than reusing `tmp`. Fixed with bounded single-assignment alias resolution applied recursively. Net ledger effect: 55/24/31/0-new — identical, but now for the right reason.

Also fixed the `parts` annotation, and set `persist-credentials: false` on this PR's job only (the other 14 are out of scope and some may need the credential).

File: scripts/check_atomic_writes.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+def _qualname(tree: ast.AST, node: ast.AST) -> str:
+    """`Class.method` rather than `method`.
+
+    The baseline key is (file, func, temp), and a BARE name lets two same-named
+    methods in one file share one row -- so a baselined clean `A._write` would
+    absorb a genuinely new leaking `C._write` and the guard would exit 0. Zero
+    live collisions today; `_write::tmp` and `_atomic_write_json::tmp` already
+    repeat across files, and `tmp` is the temp name in most rows, so it is a
+    matter of time rather than of luck."""
+    parts: list[str] = []
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `44e13e8b3`.

The substring→identity finding was the important one and I reproduced it before acting: `os.unlink(tmp_backup)` credited cleanup for temp `tmp`, so a genuinely leaking site read `CLEANS_UP`. A false clean is worse than a false flag here — the leak is invisible *and* excluded from the debt ledger, so nothing revisits it. An earlier pass raised this as a NOTE and I accepted it as documented; that was the error.

Recall was checked in the same breath, because identity alone was also wrong: it turned the false clean into a false flag at `engine.py:1824`, where the cleanup rebuilds the path rather than reusing `tmp`. Fixed with bounded single-assignment alias resolution applied recursively. Net ledger effect: 55/24/31/0-new — identical, but now for the right reason.

Also fixed the `parts` annotation, and set `persist-credentials: false` on this PR's job only (the other 14 are out of scope and some may need the credential).

File: scripts/check_atomic_writes.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+            pairs.append((names, _unparse(value)))
+
+    for names, rhs in pairs:
+        if any(m in rhs for m in _TEMP_MAKERS) or any(x in rhs for x in _TEMP_SUFFIXES):
+            temps.update(names)
+
+    for _ in range(len(pairs) + 1):  # bounded: at most one new name per pass
+        grew = False
+        for names, rhs in pairs:
+            if any(t in rhs for t in temps) and not set(names) <= temps:
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `44e13e8b3`.

The substring→identity finding was the important one and I reproduced it before acting: `os.unlink(tmp_backup)` credited cleanup for temp `tmp`, so a genuinely leaking site read `CLEANS_UP`. A false clean is worse than a false flag here — the leak is invisible *and* excluded from the debt ledger, so nothing revisits it. An earlier pass raised this as a NOTE and I accepted it as documented; that was the error.

Recall was checked in the same breath, because identity alone was also wrong: it turned the false clean into a false flag at `engine.py:1824`, where the cleanup rebuilds the path rather than reusing `tmp`. Fixed with bounded single-assignment alias resolution applied recursively. Net ledger effect: 55/24/31/0-new — identical, but now for the right reason.

Also fixed the `parts` annotation, and set `persist-credentials: false` on this PR's job only (the other 14 are out of scope and some may need the credential).

File: scripts/check_atomic_writes.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+        # A false-positive row in a debt ledger is not noise; it is a booby-trapped
+        # work item. Narrowing here TIGHTENS the allowlist rather than loosening
+        # it: a temp created in a CALLER is not claimed, which is the documented
+        # cross-function limitation, not a new hole.
+        if func is not None and not _born_in(func, temp):
+            continue
+        if func is None:
+            rows.append({"file": rel, "line": node.lineno, "func": "<module>",
+                         "temp": temp, "verdict": "NO_HANDLER"})
+            continue
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `44e13e8b3`.

The substring→identity finding was the important one and I reproduced it before acting: `os.unlink(tmp_backup)` credited cleanup for temp `tmp`, so a genuinely leaking site read `CLEANS_UP`. A false clean is worse than a false flag here — the leak is invisible *and* excluded from the debt ledger, so nothing revisits it. An earlier pass raised this as a NOTE and I accepted it as documented; that was the error.

Recall was checked in the same breath, because identity alone was also wrong: it turned the false clean into a false flag at `engine.py:1824`, where the cleanup rebuilds the path rather than reusing `tmp`. Fixed with bounded single-assignment alias resolution applied recursively. Net ledger effect: 55/24/31/0-new — identical, but now for the right reason.

Also fixed the `parts` annotation, and set `persist-credentials: false` on this PR's job only (the other 14 are out of scope and some may need the credential).

**Comment by WingedGuardian** [comment]
GLM-5.3 review applied in `ee3dfdb2a`. Its verdict was BLOCK, and most of that was a sandbox artifact — its review copy has no `scripts/hooks/`, so my ledger's rows were unresolvable there. On the real tree the guard exits 0 and this job passed CI in 9s.

But two findings survive that, and both were real:

**`Path(tmp).replace(dest)` produced no row at all.** The receiver kept its wrapper, so the temp expression was `Path(tmp_path)` — `_born_in` looks up a bare name, and that isn't one, so the site was silently dropped rather than judged. Three real files use the style (`ego/config.py:98`, `mcp/health/settings.py:706`, `outreach/config.py:231`) and all were invisible. They do clean up, so nothing was hidden — but the guard's whole claim is that a new one can't arrive silently, and in this house style it could.

**The published counts didn't reproduce** — README said 31 dirty across 28 files; it's 27. A test now derives all the counts from the tree so they can't drift apart again.

And it exposed an overclaim in my own prose: the baseline key excludes the line number, so a second unprotected write inside an already-baselined function passes silently. Qualified the claim rather than re-keying, since line-numbered keys would churn on every unrelated edit.

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
@codex review

File: scripts/check_atomic_writes.py [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+        handlers = _handlers_covering(func, node.lineno)
+        if not handlers:
+            verdict = "NO_HANDLER"
+        elif _unlinks(handlers, temp, func):
+            verdict = "CLEANS_UP"
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `bebb228b1`. Cleanup is now required on every APPLICABLE path rather than any handler node: `_can_catch_move` filters handlers that could catch a move, and the site is clean only when all of those unlink and one fully catches.

Getting there took two corrections. Requiring every COLLECTED handler to unlink flagged four genuinely clean live sites, because handlers nest — an inner `except BaseException:` that unlinks and re-raises means the outer never sees the temp. Handlers now resolve innermost-first, and a covering `finally` that unlinks dominates them. Then applicability itself had to flip from a denylist to an allowlist: one sibling `except RuntimeError:` beside a correct `except OSError:` was turning clean sites into leaks, and MEASURED, this tree has 296 multi-handler try statements and 590 handler occurrences the denylist wrongly admitted.

File: scripts/check_atomic_writes.py [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+        if func is not None and not _born_in(func, temp):
+            continue
+        if func is None:
+            rows.append({"file": rel, "line": node.lineno, "func": "<module>",
+                         "temp": temp, "verdict": "NO_HANDLER"})
+            continue
+        handlers = _handlers_covering(func, node.lineno)
+        if not handlers:
+            verdict = "NO_HANDLER"
+        elif _unlinks(handlers, temp, func):
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `bebb228b1`. The window is now the whole sequence: after a CLEANS_UP verdict, any write to the temp that no try covers downgrades it.

Two details worth recording. Creation is deliberately NOT the trigger — if `mkstemp` itself fails there is nothing on disk to leak — and counting it flagged `util/atomic.py`, the reference implementation this guard tells people to adopt, whose `mkstemp` sits outside the try while the write it protects sits inside. And the first version used a substring test against the unparsed call, which is the defect `_unlinks` already documents fixing; it now resolves operand identity, and checks EVERY write rather than the earliest, since a covered first write followed by an uncovered append was reading clean.

MEASURED before landing: zero live sites reclassified, so this is coverage rather than churn.

File: scripts/check_atomic_writes.py [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+            continue
+        # str.replace(old, new) takes >=2 args; Path.replace(target) exactly one;
+        # os.replace(src, dst) two but is named unambiguously.
+        if owner != "os" and node.func.attr != "move" and len(node.args) != 1:
+            continue
+        # A kwarg-only call (`os.replace(src=a, dst=b)`) has no positional args;
+        # falling through would set temp = "os" and then credit ANY os.unlink in
+        # a handler as cleanup.
+        if owner in ("os", "shutil") and len(node.args) < 2:
+            continue
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `0656db705`. `_operand(call, index, *names)` resolves both operands by position OR keyword, so `os.replace(src=tmp, dst=target)` is read rather than dropped. The kwarg-only spelling previously fell through to `temp = "os"`, which credited any `os.unlink` in a handler as cleanup for it.

File: scripts/check_atomic_writes.py [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+    tree = ast.parse(src)
+    rows: list[dict] = []
+    for node in ast.walk(tree):
+        if not isinstance(node, ast.Call):
+            continue
+        # A filesystem move is ALWAYS an attribute call. A bare `replace(x, ...)`
+        # is dataclasses.replace -- the largest false-positive class this guard
+        # had, and nothing about it touches a filesystem.
+        if not isinstance(node.func, ast.Attribute):
+            continue
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `0656db705`. `_directly_imported_moves(tree)` allowlists the names an `os`/`shutil` import actually binds, so `from os import replace` is recognised while `dataclasses.replace` stays excluded — dropping the attribute rule wholesale would have reopened the largest false-positive class this guard has had. Module-level imports only, and the call site pairs it with `_locally_rebound` so a function that shadows the name does not inherit the meaning.

File: scripts/check_atomic_writes.py [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+        # a quarantined corrupt entry kept as evidence (guardian/alert/queue.py).
+        # A false-positive row in a debt ledger is not noise; it is a booby-trapped
+        # work item. Narrowing here TIGHTENS the allowlist rather than loosening
+        # it: a temp created in a CALLER is not claimed, which is the documented
+        # cross-function limitation, not a new hole.
+        if func is not None and not _born_in(func, temp):
+            continue
+        if func is None:
+            rows.append({"file": rel, "line": node.lineno, "func": "<module>",
+                         "temp": temp, "verdict": "NO_HANDLER"})
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `bebb228b1`. Module scope is now a scope: `_born_in` and `_handlers_covering` both take any AST node, so the module tree serves as the enclosing scope instead of the old unconditional NO_HANDLER stamp.

It was wrong in both directions at once, as you and CodeRabbit both noted — a module-level move-aside produced a dirty row for a DURABLE operand while a module-level write that does clean up still read dirty. Folding it into the normal path also surfaced a latent crash: those rows then reached `_qualname(tree, None)`, an AttributeError that no live file triggers today.

File: scripts/check_atomic_writes.py [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+        merely named `tmp_dir_listing` does not, and a substring test cannot tell
+        them apart."""
+        for n in ast.walk(value):
+            if isinstance(n, ast.Call):
+                fn = getattr(n.func, "attr", None) or getattr(n.func, "id", None)
+                if fn in _TEMP_MAKERS:
+                    return True
+            if (isinstance(n, ast.Constant) and isinstance(n.value, str)
+                    and n.value.endswith(_TEMP_SUFFIXES)):
+                return True
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py:431]: Fixed. The suffix test is relaxed for f-string pieces, so a marker followed by a unique component (`f".{name}.restore-tmp-{getpid()}"`) is recognised. That site had produced NO ROW AT ALL, which is why the error surfaced in the published denominator rather than in any verdict.

`bebb228b1` narrows it further: the marker must describe the path being ASSIGNED, since walking every descendant literal made `load_record(excluded_suffix='.tmp')` mark its target a temp. Module-level string constants are also resolved now — `f"{stem}{_STAGING_SUFFIX}"` carries no `ast.Constant` piece at all — gated at the use site by `_locally_rebound`.

File: scripts/check_atomic_writes.py [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+_TRANSPARENT = ("Path", "str", "os.fspath", "pathlib.Path")
+
+
+def _strip_wrappers(expr: ast.AST) -> ast.AST:
+    """Remove path-neutral wrappers: Path(x), str(x), x.expanduser()/resolve()."""
+    node = expr
+    for _ in range(8):  # bounded; nesting deeper than this is not real code
+        if isinstance(node, ast.Call) and _unparse(node.func) in _TRANSPARENT and node.args:
+            node = node.args[0]
+            continue
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed. `_strip_wrappers` unwraps only single-argument transparent wrappers; a multi-argument `Path(tmpdir, "payload")` names a CHILD and is left intact, so the site is UNMATCHED rather than wrongly cleared — reducing it to its first argument had recorded the DIRECTORY as the temp, letting a handler that removed the directory read as cleaning up a child it never touched.

`bebb228b1` applies the same arity rule on the marker side: `Path(...)`/`str(...)` are transparent at one argument and opaque beyond it.

File: scripts/check_atomic_writes.py [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+    want_node = ast.parse(temp_expr, mode="eval").body
+    wants = {_core_name(want_node), _resolve(want_node, aliases)}
+    for grp in nodes:
+        for item in (grp if isinstance(grp, list) else [grp]):
+            for n in ast.walk(item):
+                if not isinstance(n, ast.Call):
+                    continue
+                fn = getattr(n.func, "attr", None) or getattr(n.func, "id", None)
+                if fn not in ("unlink", "remove"):
+                    continue
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py:774]: Fixed in `bebb228b1`. `remove` credits cleanup only when the owner resolves to `os`/`shutil`; a list's `.remove(x)` drops an element and was crediting a file delete, which made a genuinely leaking site read CLEANS_UP — invisible AND off the ledger. `unlink` needs no such test, since no builtin container has one. A bare `from os import remove` is deliberately not credited: that costs a false flag, which is the direction this function is allowed to be wrong in.

File: scripts/check_atomic_writes.py [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+    whose right-hand side mentions a name already known to be a temp. Anything
+    still unreached -- a parameter, a field, a path built from user input -- is
+    not a temp, and the site is not this guard's business.
+    """
+    if not temp_expr:
+        return False
+    root = temp_expr.split(".")[0].split("[")[0].strip()
+
+    def _bound_names(target) -> list[str]:
+        return [n.id for n in ast.walk(target) if isinstance(n, ast.Name)]
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `5226e1baf`. `_bound_names` binds whole attribute and subscript paths instead of walking to every `ast.Name`, so `self.staging = path.with_suffix('.tmp')` no longer records **`self`** as a temp — which had made every later `self.<anything>.replace(dst)` share that root and report as a leak, carrying the unlink remediation at a durable file. That is the durable-operand trap the born-here rule exists to prevent, re-entered through the binding side.

File: scripts/check_atomic_writes.py [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+def key(row: dict) -> str:
+    """Baseline identity. Deliberately EXCLUDES the line number, which shifts on
+    every unrelated edit above it and would turn the ledger into churn."""
+    return f"{row['file']}::{row['func']}::{row['temp']}"
+
+
+def scan(repo: Path) -> tuple[list[dict], list[str]]:
+    rows, errors = [], []
+    for base in ("src", "scripts"):
+        for path in sorted((repo / base).rglob("*.py")):
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `bebb228b1`. `az_plugins/` joins `SCAN_ROOTS`, and `UNSCANNED_ROOTS` records `tests/` as a deliberate exclusion with its reason. MEASURED: az_plugins holds 11 modules and produces zero rows, so the gap was coverage rather than verdicts.

A hardcoded root list fails for the reason this one did — the tree did not exist when the list was written — so a test now pins that every top-level Python-bearing tree is either scanned or documented as excluded. Its own limits (dot-directories, symlinked subtrees) are recorded in the guard's blind-spot list.

File: scripts/check_atomic_writes.py [hunk 15]
```diff
@@ -0,0 +1,617 @@
...
+            return [target.id]
+        if isinstance(target, (ast.Attribute, ast.Subscript)):
+            return [_unparse(target)]
+        if isinstance(target, ast.Starred):
+            return _bound_names(target.value)
+        return []
+
+    temps: set[str] = set()
+    # Seed: `with tempfile.NamedTemporaryFile(...) as tmp:` and friends.
+    for n in ast.walk(func):
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `bebb228b1`. Origin tracking now respects statement order and scope: bindings count only when they reach the move (`lineno <= before_lineno`), and `_own_scope` does not descend into a nested `def`, since a binding in another scope is a different variable.

Textual order only approximates reaching-definitions — a loop can execute a later line first — and it is the safe approximation, since it can only drop a row, never invent one. Both shapes are pinned as fixtures.

File: scripts/check_atomic_writes.py [hunk 16]
```diff
@@ -0,0 +1,617 @@
...
+            for n in ast.walk(item):
+                if not isinstance(n, ast.Call):
+                    continue
+                fn = getattr(n.func, "attr", None) or getattr(n.func, "id", None)
+                if fn not in ("unlink", "remove"):
+                    continue
+                targets: set[str] = set()
+                for a in n.args:
+                    targets.add(_core_name(a))
+                    targets.add(_resolve(a, aliases))
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py:792]: Fixed in `bebb228b1`. `_unlinks` now reads keyword targets, so `os.unlink(path=tmp)` credits cleanup exactly as the positional spelling does. Reading only `n.args` had reported the PROTECTED site as LEAKS — a false flag, whose cost is sending someone to fix code that is already correct.

File: scripts/check_atomic_writes.py [hunk 17]
```diff
@@ -0,0 +1,617 @@
...
+        relaxation is scoped to JoinedStr pieces rather than applied globally.
+        """
+        for n in ast.walk(value):
+            if isinstance(n, ast.Call):
+                fn = getattr(n.func, "attr", None) or getattr(n.func, "id", None)
+                if fn in _TEMP_MAKERS:
+                    return True
+            if (isinstance(n, ast.Constant) and isinstance(n.value, str)
+                    and n.value.endswith(_TEMP_SUFFIXES)):
+                return True
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py:431]: Fixed in `bebb228b1`. `_makes_temp` descends only into calls that BUILD a path (`_PATH_BUILDERS`, `_TEMP_MAKERS`, and single-argument `Path`/`str` wrappers), so a constant buried in an unrelated call's arguments no longer marks its target a temp.

The first version of this fix was too aggressive and an audit caught it: omitting `Path`/`str` made `tmp = Path(str(p) + '.tmp')` and even `tmp = Path(mkstemp()[1])` produce no row at all — invisible rather than misjudged, a regression the previous code did not have. Transparency is now gated on ARITY, so the single-argument form keeps its row while the multi-argument join stays opaque.

File: scripts/check_atomic_writes.py [hunk 18]
```diff
@@ -0,0 +1,617 @@
...
+def _handlers_covering(func: ast.AST, lineno: int) -> list:
+    """Every except-handler and finally-body whose TRY BODY contains lineno.
+
+    Deliberately the try BODY, not the whole Try node: a rename sitting inside an
+    `except:` block is not protected by that same handler."""
+    out: list = []
+    for n in ast.walk(func):
+        if isinstance(n, ast.Try) and any(
+            c.lineno <= lineno <= (c.end_lineno or c.lineno) for c in n.body
+        ):
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `bebb228b1`. `_handlers_covering` now applies the finalizer to the body, the handlers and the `else` suite alike, while `except` handlers stay limited to what the body raises. `ast.TryStar` is included — it was invisible to an isinstance test naming only `ast.Try`, and the suite contained zero `except*` anywhere, so that widening would have shipped unpinned. Both have fixtures now.

File: scripts/check_atomic_writes.py [hunk 19]
```diff
@@ -0,0 +1,817 @@
...
+        # those shapes -- and this guard prints "unlink the temp" as the
+        # remediation, so following it would have deleted a live credential
+        # (guardian/cred_integrity.py), a user's file (dashboard/routes/files.py),
+        # pending telemetry on its restore path (observability/span_ingest.py) and
+        # a quarantined corrupt entry kept as evidence (guardian/alert/queue.py).
+        # A false-positive row in a debt ledger is not noise; it is a booby-trapped
+        # work item. Narrowing here TIGHTENS the allowlist rather than loosening
+        # it: a temp created in a CALLER is not claimed, which is the documented
+        # cross-function limitation, not a new hole.
+        if func is not None and not _born_in(func, temp, module_consts):
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/check_atomic_writes.py]: Fixed in `bebb228b1` — and thank you for finding this independently of Codex's `:426`, which is what moved it from a known limitation to a defect.

`scope = func if func is not None else tree` is exactly the fix you proposed: both `_born_in` and `_handlers_covering` accept any AST node, so the module tree serves as the enclosing scope. Both consequences you named are now correct — a module-level move-aside on a durable operand produces no row, and a module-level atomic write that unlinks in `except`/`finally` reads CLEANS_UP. Verified in all three directions (durable, clean, leaking) as fixtures.

One thing the fix surfaced: those rows then reached `_qualname(tree, None)`, which dereferences `node.name` — an AttributeError that no live file triggers, since the tree has no module-level atomic write today. The old early-return had been hiding it.

**Comment by WingedGuardian** [comment]
@codex review
